File: tests/sandbox.test.ts
import { describe, it, expect, beforeAll } from 'vitest';
import { loadEvents } from '../src/batch/runner';
import {
  runSandbox,
  runScenario,
  countLockedViolations,
  SandboxReport,
} from '../src/policy/sandbox';
import { DEFAULT_TUNABLES } from '../src/policy/config';
import { getTunables, resetTunables } from '../src/decisions/rules';

const now = Date.UTC(2026, 8, 1, 14, 30);

describe('backtest sandbox', () => {
  const events = loadEvents().sort((a, b) => a.eventId.localeCompare(b.eventId));
  let report: SandboxReport;

  beforeAll(() => {
    report = runSandbox(events, { now });
  });

  it('always restores defaults after a scenario (no leaked tunables)', () => {
    resetTunables();
    expect(getTunables()).toEqual(DEFAULT_TUNABLES);
  });

  it('returns a baseline plus the swept scenarios', () => {
    expect(report.scenarios.length).toBeGreaterThanOrEqual(3);
    expect(report.baseline.recoveryRate).toBeGreaterThan(0);
  });

  it('locked safety/compliance rules never violate in any scenario', () => {
    expect(report.lockedInvariant).toBe(true);
    for (const s of report.scenarios) {
      const total = Object.values(s.lockedViolations).reduce((a, b) => a + b, 0);
      expect(total).toBe(0);
    }
  });

  it('tunables actually move recovery vs conservative tail', () => {
    const aggressive = report.scenarios.find((s) => s.scenarioLabel === 'aggressive');
    const conservative = report.scenarios.find((s) => s.scenarioLabel === 'conservative');
    expect(aggressive).toBeTruthy();
    expect(conservative).toBeTruthy();
    // Aggressive config grants more retries -> at least as much recovery & touches
    expect(aggressive!.metrics.touchesSent).toBeGreaterThanOrEqual(conservative!.metrics.touchesSent);
  });

  it('produces a headline mentioning the locked-rule invariant', () => {
    expect(report.headline).toContain('locked');
  });

  it('countLockedViolations reports zero even on extreme aggression', () => {
    const res = runScenario(
      events,
      now,
      'extreme',
      { maxRetryAttempts: 20, maxTouchesPerCustomer: 20, maxVoiceCalls: 20 },
      'data/test.sandbox.extreme.log.jsonl',
    );
    const total = Object.values(res.lockedViolations).reduce((a, b) => a + b, 0);
    expect(total).toBe(0);
  });
});

File: go/rz/sandbox_test.go
package rz

import (
	"strings"
	"testing"
)

func TestSandboxRestoresDefaults(t *testing.T) {
	ResetTunables()
	got := GetTunables()
	want := DefaultTunables()
	if got != want {
		t.Fatalf("tunables not restored: %+v vs %+v", got, want)
	}
}

func TestSandboxReport(t *testing.T) {
	events := testEvents(t)
	now := BatchNow()

	report, err := RunSandbox(events, now)
	if err != nil {
		t.Fatal(err)
	}
	if len(report.Scenarios) < 3 {
		t.Fatalf("scenarios=%d, want >=3", len(report.Scenarios))
	}
	if report.Baseline.RecoveryRate <= 0 {
		t.Fatal("baseline recoveryRate <= 0")
	}
}

func TestSandboxLockedInvariant(t *testing.T) {
	events := testEvents(t)
	now := BatchNow()

	report, err := RunSandbox(events, now)
	if err != nil {
		t.Fatal(err)
	}
	if !report.LockedInvariant {
		t.Fatal("lockedInvariant should be true")
	}
	for _, s := range report.Scenarios {
		if s.LockedViolations.Sum() != 0 {
			t.Fatalf("scenario %s has locked violations: %+v", s.ScenarioLabel, s.LockedViolations)
		}
	}
}

func TestSandboxTunablesMoveRecovery(t *testing.T) {
	events := testEvents(t)
	now := BatchNow()

	report, err := RunSandbox(events, now)
	if err != nil {
		t.Fatal(err)
	}
	var aggressive, conservative *SandboxResult
	for i := range report.Scenarios {
		switch report.Scenarios[i].ScenarioLabel {
		case "aggressive":
			aggressive = &report.Scenarios[i]
		case "conservative":
			conservative = &report.Scenarios[i]
		}
	}
	if aggressive == nil || conservative == nil {
		t.Fatal("missing aggressive/conservative scenarios")
	}
	if aggressive.Metrics.TouchesSent < conservative.Metrics.TouchesSent {
		t.Fatalf("aggressive touches=%d < conservative touches=%d",
			aggressive.Metrics.TouchesSent, conservative.Metrics.TouchesSent)
	}
}

func TestSandboxHeadline(t *testing.T) {
	report, err := RunSandbox(testEvents(t), BatchNow())
	if err != nil {
		t.Fatal(err)
	}
	if !strings.Contains(report.Headline, "locked") {
		t.Fatal("headline should mention the locked-rule invariant")
	}
}

func TestSandboxExtremeScenario(t *testing.T) {
	events := testEvents(t)
	res, err := RunScenario(events, BatchNow(), "extreme",
		func(tc *TunableConfig) {
			tc.MaxRetryAttempts = 20
			tc.MaxTouchesPerCustomer = 20
			tc.MaxVoiceCalls = 20
		},
		"data/test.sandbox.extreme.log.jsonl")
	if err != nil {
		t.Fatal(err)
	}
	if res.LockedViolations.Sum() != 0 {
		t.Fatalf("extreme scenario locked violations: %+v", res.LockedViolations)
	}
}
